HPCC-34942 Tool to analyze HPCC running on aks

shamser opened the pull request:
Cost optimization and resource analysis tool for Azure Kubernetes Service clusters running HPCC workloads.

**Key Features**
Resource Analysis: Real-time CPU/memory utilization tracking with historical idle pattern detection
Cost Optimization: Estimates monthly costs, identifies waste, calculates potential savings scenarios
Workload Integration: Connects to HPCC ECL Watch for job-aware resource analysis
Automated Recommendations: Generates actionable optimization suggestions (right-sizing, autoscaling, consolidation)
Waste Detection: Identifies idle nodes, underutilized resources, and misplaced agent workloads

**Output**
Cost estimates with enterprise discounts
Node-by-node utilization analysis
Prioritized optimization opportunities
Integration with Azure Cost Management APIs when available

**Primary Use Cases**
Cost Control - Quantify and reduce cloud spending
Right-sizing - Match resources to actual workload demands
Efficiency Monitoring - Track cluster performance over time
Capacity Optimization - Balance performance with cost efficiency

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

ghalliday reviewed (changes requested):
The script looks reasonable - I haven't checked details.  Copilot's comments are relevant about catest.  Otherwise it shoudl be fine to merge.
Did you mean to assign jake/ken rather than add for review?

shamser commented:
> The script looks reasonable - I haven't checked details. Copilot's comments are relevant about catest. Otherwise it shoudl be fine to merge. Did you mean to assign jake/ken rather than add for review?

yes, the assignment was a mistake.

jakesmith commented on tools/hpcc_aks_util.pl:
```diff
@@ -0,0 +1,772 @@
+  --show-pods                           Show pods in the output
+  --show-nodes                          Show detailed node information with pods per node
+  --ea-discount PERCENT                 Enterprise Agreement discount percentage (default: 10)
+  --ds_v4_family_discount PERCENT       Dds_v4 family discount percentage for D16ds v4 and larger (default: 10)
```
fine for now, but would be good to generalize this to something like
--discountnode "<vmtype>=<pct>"

jakesmith commented on tools/hpcc_aks_util.pl:
```diff
@@ -0,0 +1,772 @@
+--ea-discount           Applied to most VM series (D-series v5, v4 smaller, E-series, F-series, B-series)
+--ds_v4_family_discount Applied specifically to D16ds v4 and larger instances
+
+Default discounts are 10% each. Costs are estimates and actual pricing may vary.
```
would it better to not default to 10% discount like this. Default to published costs, unless --discount's apply?

jakesmith commented on tools/hpcc_aks_util.pl:
```diff
@@ -0,0 +1,772 @@
+my $show_pods = 0;
+my $show_nodes = 0;
+my $ea_discount = 10;
+my $ds_v4_family_discount = 10;
```
see prev. comment - wouldn't it be better not to apply default [wrong] discounts, i.e. better for these values to be 0?

jakesmith commented on tools/hpcc_aks_util.pl:
```diff
@@ -0,0 +1,772 @@
+    
+    my %vm_costs = (
+        # Standard D-series v5 (EA discount applied)
+        'Standard_D2s_v5' => 0.096 * $ea_multiplier,
```
future: these hard-coded figures should be in a table somewhere, ideally in an external file, so the script doesn't need updaating

jakesmith commented on tools/hpcc_aks_util.pl:
```diff
@@ -0,0 +1,772 @@
+    if (exists $vm_costs{$vm_size}) {
+        return sprintf("%.3f", $vm_costs{$vm_size});
+    } else {
+        # Default fallback - estimate based on CPU count if available
```
if this happens, won't it be very misleading? i.e. silently defaulting figures if an unknown VM type is found.
I think it would be better to error, with the message saying the vmcosts table needs updating.

jakesmith commented on tools/hpcc_aks_util.pl:
```diff
@@ -0,0 +1,772 @@
+    
+    if ($mem_raw =~ /^(\d+(?:\.\d+)?)Ki$/) {
+        return $1 / 1024;  # Convert to Mi
+    } elsif ($mem_raw =~ /^(\d+(?:\.\d+)?)Mi$/) {
```
doesn't handle gigabyte notations

jakesmith commented on tools/hpcc_aks_util.pl:
```diff
@@ -0,0 +1,772 @@
+}
+
+# Get credentials for the AKS cluster
+execute_command("az aks get-credentials --resource-group '$resource_group' --name '$cluster_name' --overwrite-existing");
```
could use --file to avoid overrwriting user config

jakesmith reviewed (commented):
@shamser - haven't gone through it all, started to add review comments on the PERL version..

Why perl?
This would be more maintainable if just a python script.

shamser committed:
HPCC-34942 Tool to analyze HPCC running on AKS

Cost optimization and resource analysis tool for Azure Kubernetes Service clusters running HPCC workloads.

Key Features
Resource Analysis: Real-time CPU/memory utilization tracking with historical idle pattern detection
Cost Optimization: Estimates monthly costs, identifies waste, calculates potential savings scenarios
Workload Integration: Connects to HPCC ECL Watch for job-aware resource analysis
Automated Recommendations: Generates actionable optimization suggestions (right-sizing, autoscaling, consolidation)
Waste Detection: Identifies idle nodes, underutilized resources, and misplaced agent workloads

Output
Cost estimates with enterprise discounts
Node-by-node utilization analysis
Prioritized optimization opportunities
Integration with Azure Cost Management APIs when available

Primary Use Cases
Cost Control - Quantify and reduce cloud spending
Right-sizing - Match resources to actual workload demands
Efficiency Monitoring - Track cluster performance over time
Capacity Optimization - Balance performance with cost efficiency

Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexisrisk.com>

ghalliday reviewed (approved):
Interesting use of emojis.
I don't have any reason not to merge.  It would be good to have a sample output included in the jira, or somewhere else.

jakesmith commented on tools/azure/hpcc-aks-cluster-insights.py:
```diff
@@ -0,0 +1,4251 @@
+                    description = f"Thor worker cluster ({workers} workers estimated)" if workers > 0 else "Thor worker cluster"
+                elif "l" in pool_name:
+                    thor_cluster = f"thor-{pool_name}"
+                    purpose = "Thor Spill Workers"
```
generally, what does "Thor Spill Workers" mean? (sounds like it might be something specific to prod400 env.)

The tool should be kept agnostic, so we can point at healthcar, fcra etc.. / any env.

jakesmith reviewed (approved):
@shamser - started to add some comments, but there is no reason to hold this back.

It is very big, and consequently hard to review.
I'd prefer it , if it was a set of discrete separate scripts if possible (could include comomn functions from a common file). That would make it easier to digest and maintain.

I think this will provide more insights, but new scripts should look to use Azure's published tables, rather than interacting with kubectl and az to pull "live" current snapshots, e.g. KubePodInventory, KubeNodeInventory, and pull the helm configuration (values yaml) if necessasry e.g. to know # of workers, the list of targets, etc.

jakesmith commented:
@shamser - can you close this?
